webpack/webpack-workspace-resolve-plugin/src/KnownDescriptionFilePlugin.ts

```diff
@@ -0,0 +1,117 @@
+    }
+
+    const target: ReturnType<Resolver['ensureHook']> = resolver.ensureHook(this.target);
+    const { fileSystem } = resolver;
+    function readDescriptionFileAsJson(
+      descriptionFilePath: string,
+      callback: (err: Error | null | undefined, data?: object) => void
+    ): void {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      fileSystem.readJson!(descriptionFilePath, callback);
+    }
+
+    function readDescriptionFileWithParse(
+      descriptionFilePath: string,
+      callback: (err: Error | null | undefined, data?: object) => void
+    ): void {
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      fileSystem.readFile(descriptionFilePath, (err: Error | null | undefined, data?: string | Buffer) => {
+        if (!data?.length) {
+          return callback(err);
+        }
+        // eslint-disable-next-line @rushstack/no-new-null
+        callback(null, JSON.parse(data.toString()));
+      });
+    }
+
+    const readDescriptionFile: (
+      descriptionFilePath: string,
+      cb: (err: Error | null | undefined, data?: object) => void
+    ) => void = fileSystem.readJson ? readDescriptionFileAsJson : readDescriptionFileWithParse;
+
+    resolver
+      .getHook(this.source)
+      .tapAsync(KnownDescriptionFilePlugin.name, (request, resolveContext, callback) => {
+        const { path } = request;
+        // No request, nothing to do.
+        if (!path) return callback();
+
+        const cache: WorkspaceLayoutCache = this._cache;
+
+        const match: IPrefixMatch<IResolveContext> | undefined =
+          cache.contextLookup.findLongestPrefixMatch(path);
+        // No description file available, proceed without.
+        if (!match) return callback();
+
+        const relativePath: string = `.${cache.normalizeToSlash(path.slice(match.index))}`;
+        const descriptionFileRoot: string = `${path.slice(0, match.index)}`;
+        const descriptionFilePath: string = `${descriptionFileRoot}${cache.resolverPathSeparator}package.json`;
+
+        const { contextForPackage } = cache;
+
+        readDescriptionFile(descriptionFilePath, (err, descriptionFileData) => {
+          if (!descriptionFileData) {
+            resolveContext.missingDependencies?.add(descriptionFilePath);
+            return callback(err);
+          }
+
+          resolveContext.fileDependencies?.add(descriptionFilePath);
+          // Store the resolver context since a WeakMap lookup is cheaper than walking the tree again
+          contextForPackage.set(descriptionFileData, match.value);
+
+          // Since we don't allow any alternative processing of request, we can mutate it
+          // instead of cloning it.
+          request.descriptionFileRoot = descriptionFileRoot;
+          request.descriptionFilePath = descriptionFilePath;
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          request.descriptionFileData = descriptionFileData as any;
```

Comment:
What type does `request.descriptionFileData` want?

Reply:
Apparently `object` now. It had been complaining in earlier iterations.